File: src/leetcode/2025/12/Solution257.cpp
#include <vector>
#include <string>
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution257
{
public:
    vector<string> binaryTreePaths(TreeNode *root)
    {
        vector<string> result;
        if (root == nullptr)
        {
            return result;
        }
        string path;
        dfs(root, path, result);
        return result;
    }

private:
    void dfs(TreeNode *node, string path, vector<string> &result)
    {
        path += to_string(node->val);
        if (node->left == nullptr && node->right == nullptr)
        {
            result.push_back(path);
            return;
        }
        path += "->";
        if (node->left != nullptr)
        {
            dfs(node->left, path, result);
        }
        if (node->right != nullptr)
        {
            dfs(node->right, path, result);
        }
    }
};

int main()
{
    Solution257 sol;

    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->right = new TreeNode(5);

    vector<string> paths = sol.binaryTreePaths(root);
    for (const string &path : paths)
    {
        cout << path << endl;
    }

    return 0;
}

File: src/leetcode/2025/12/Solution257.java
import java.util.ArrayList;
import java.util.List;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution257 {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> answer = new ArrayList<>();

        if (root != null)
            searchBT(root, "", answer);

        return answer;
    }

    private void searchBT(TreeNode root, String path, List<String> answer) {
        if (root.left == null && root.right == null)
            answer.add(path + root.val);
        if (root.left != null)
            searchBT(root.left, path + root.val + "->", answer);
        if (root.right != null)
            searchBT(root.right, path + root.val + "->", answer);
    }

    public static void main(String[] args) {
        Solution257 sol = new Solution257();

        TreeNode root = new TreeNode(1, new TreeNode(2, null, new TreeNode(5)), new TreeNode(3));
        List<String> answer = sol.binaryTreePaths(root);
        System.out.println(answer);
    }
}
